src/main/java/com/google/api/codegen/transformer/py/PythonImportSectionTransformer.java

```diff
@@ -332,27 +335,57 @@ private ImportFileView createImport(String moduleName, String attributeName, Str
 
   public ImportSectionView generateTypesImportSection(
       Model model, GapicProductConfig productConfig) {
-    return ImportSectionView.newBuilder()
-        .appImports(generateTypesProtoImports(model, productConfig))
+    ImportSectionView.Builder importView = generateTypesProtoImports(model, productConfig);
+    return importView
         .externalImports(generateTypesExternalImports())
         .standardImports(generateTypesStandardImports())
         .build();
   }
 
-  private List<ImportFileView> generateTypesProtoImports(
+  private ImportSectionView.Builder generateTypesProtoImports(
       Model model, GapicProductConfig productConfig) {
-    ModelTypeTable typeTable = emptyTypeTable(productConfig);
-    Set<ImportFileView> imports = new TreeSet<>(importFileViewComparator());
+    ImportSectionView.Builder importView = ImportSectionView.newBuilder();
+
+    ModelTypeTable localTypeTable = emptyTypeTable(productConfig);
+    ModelTypeTable sharedTypeTable = emptyTypeTable(productConfig);
+    ModelTypeTable allTypeTable = emptyTypeTable(productConfig);
+
+    // Imports from the same API client library package.
+    Set<ImportFileView> localImports = new TreeSet<>(importFileViewComparator());
+
+    // Shared imports, e.g. protobuf imports.
+    Set<ImportFileView> sharedImports = new TreeSet<>(importFileViewComparator());
+
+    // All imports.
+    Set<ImportFileView> appImports = new TreeSet<>(importFileViewComparator());
+    Map<String, IMPORT_TYPE> importNamesAndTypes = new HashMap<>();
+
+    List<ProtoElement> elements = Lists.newArrayList(model.getRoots());
+    String serviceFullName = elements.get(0).getFullName();
+    String packageName = serviceFullName.substring(0, serviceFullName.lastIndexOf("."));
 
     // Save proto file import names to the type table for disambiguation.
     List<ProtoFile> protoFileDependencies = model.getFiles();
-    populateTypeTable(protoFileDependencies, typeTable);
+    populateTypeTables(protoFileDependencies, allTypeTable, importNamesAndTypes, packageName);
 
     // Get disambiguated imports.
-    for (Map.Entry<String, TypeAlias> entry : typeTable.getImports().entrySet()) {
-      imports.add(generateAppImport(entry.getKey(), entry.getValue().getNickname()));
+    for (Map.Entry<String, TypeAlias> entry : allTypeTable.getImports().entrySet()) {
+      String importFullName = entry.getKey();
+      appImports.add(generateAppImport(importFullName, entry.getValue().getNickname()));
+      switch (importNamesAndTypes.get(importFullName)) {
+        case LOCAL:
+          localImports.add(generateAppImport(importFullName, entry.getValue().getNickname()));
```

Comment:
I think the logic used here is more complex than it needs to be. Something like this might be simpler?

- Make `populateTypeTables` take two sets, local and shared.
- `populateTypeTables` just puts the import in the right place, no need for `importNamesAndTypes`.
- then just `appImports = ImmutableSortedSet.builder().addAll(shared).addAll(local).build().asList()`